cmd/minikube/cmd/config/profile_list.go

```diff
@@ -56,98 +58,111 @@ var profileListCmd = &cobra.Command{
+	cp, err := config.PrimaryControlPlane(p.Config)
+	if err != nil {
+		exit.WithError("error getting primary control plane", err)
+	}
 
+	host, err := machine.LoadHost(api, driver.MachineName(*p.Config, cp))
+	if err != nil {
+		return ""
 	}
 
+	cr, err := machine.CommandRunner(host)
 	if err != nil {
-		glog.Warningf("error loading profiles: %v", err)
+		return ""
 	}
 
-}
+	hostname, _, port, err := driver.ControlPlaneEndpoint(p.Config, &cp, host.DriverName)
+	if err != nil {
+		return ""
+	}
 
-var printProfilesJSON = func() {
-	api, err := machine.NewAPIClient()
+	status, err := kverify.APIServerStatus(cr, hostname, port)
 	if err != nil {
-		glog.Errorf("failed to get machine api client %v", err)
+		glog.Warningf("error getting apiserver status for %s: %v", p.Name, err)
+		return ""
 	}
-	defer api.Close()
+	return status.String()
+}
 
-	validProfiles, invalidProfiles, err := config.ListProfiles()
-	for _, v := range validProfiles {
-		cp, err := config.PrimaryControlPlane(v.Config)
+func renderProfilesTable(ps [][]string) {
+	table := tablewriter.NewWriter(os.Stdout)
+	table.SetHeader([]string{"Profile", "VM Driver", "Runtime", "IP", "Port", "Version", "Status"})
+	table.SetAutoFormatHeaders(false)
+	table.SetBorders(tablewriter.Border{Left: true, Top: true, Right: true, Bottom: true})
+	table.SetCenterSeparator("|")
+	table.AppendBulk(ps)
+	table.Render()
+}
+
+func profilesToTableData(profiles []*config.Profile) [][]string {
+	var data [][]string
+	for _, p := range profiles {
+		cp, err := config.PrimaryControlPlane(p.Config)
 		if err != nil {
 			exit.WithError("error getting primary control plane", err)
 		}
-		status, err := machine.Status(api, driver.MachineName(*v.Config, cp))
-		if err != nil {
-			glog.Warningf("error getting host status for %s: %v", v.Name, err)
-		}
-		v.Status = status
+
+		data = append(data, []string{p.Name, p.Config.Driver, p.Config.KubernetesConfig.ContainerRuntime, cp.IP, strconv.Itoa(cp.Port), p.Config.KubernetesConfig.KubernetesVersion, p.Status})
 	}
+	return data
+}
 
-	var valid []*config.Profile
-	var invalid []*config.Profile
+func warnInvalidProfiles(invalidProfiles []*config.Profile) {
+	if invalidProfiles == nil {
+		return
+	}
 
-	if validProfiles != nil {
-		valid = validProfiles
-	} else {
-		valid = []*config.Profile{}
+	out.WarningT("Found {{.number}} invalid profile(s) ! ", out.V{"number": len(invalidProfiles)})
+	for _, p := range invalidProfiles {
+		out.ErrT(out.Empty, "\t "+p.Name)
 	}
 
-	if invalidProfiles != nil {
-		invalid = invalidProfiles
-	} else {
-		invalid = []*config.Profile{}
+	out.ErrT(out.Tip, "You can delete them using the following command(s): ")
+	for _, p := range invalidProfiles {
+		out.Err(fmt.Sprintf("\t $ minikube delete -p %s \n", p.Name))
 	}
+}
 
-	var body = map[string]interface{}{}
+func printProfilesJSON() {
+	validProfiles, invalidProfiles, err := config.ListProfiles()
 
+	updateProfilesStatus(validProfiles)
+
+	var body = map[string]interface{}{}
```

Comment:
I see `body`  `contains only []*config.Profile`.
is it possible to replace interface here with exact type?

Reply:
About this one, we are assigning `body["erorr"] = err` a couple of lines after.
That's why I kept it this one.